### PR TITLE
[4.0][WIP] Field password as Custom Element

### PR DESCRIPTION
**Pull request by dgrammatiko**
Part of the effort to convert all fields to Custom Elements

### Summary of Changes
- One script to rule them all
- Ability to set multiple validation handlers (per field)
- Remove Bootstrap/Font Awesome and other unneeded dependencies so the field can play nicely in every known css framework
-  Some code improvements

@ciar4n can you take care of the css @todos ?

### Testing Instructions

For easy testing you can apply this patch and check the admin login:
screen shot 2017-12-30 at 17 47 02
screen shot 2017-12-30 at 17 47 12
screen shot 2017-12-30 at 17 47 20
screen shot 2017-12-30 at 17 47 28

**Issues**
- The show/hide button doesn't look like a button
- Still Bootstrap specific
- Still Font awesome specific
- Needs A11Y audit 
- Needs final approval form the design team

But the functionality should be spot on!

### CODE REVIEW

File: administrator/language/en-GB/en-GB.mod_login.ini [hunk 1]
```diff
@@ -9,3 +9,5 @@ MOD_LOGIN_LOGIN_TITLE="Administrator Log In"
 MOD_LOGIN_XML_DESCRIPTION="This module displays a username and password login form. It should not be unpublished."
 MOD_LOGIN_REMIND="Forgot your username?"
 MOD_LOGIN_RESET="Forgot your password?"
+JFIELD_PASSWORD_INDICATE_INCOMPLETE="Incomplete or something similar?"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.mod_login.ini]: ????

File: administrator/modules/mod_login/tmpl/default.php [hunk 2]
```diff
@@ -42,21 +44,23 @@ class="form-control input-full"
...
-				<span class="input-group-addon">
-					<span class="fa fa-eye" aria-hidden="true"></span>
-					<span class="sr-only"><?php echo Text::_('JSHOW'); ?></span>
-				</span>
-			</div>
+			<joomla-field-password reveal="true" min-length="2">
+				<label for="mod-login-password"><?php echo Text::_('JGLOBAL_PASSWORD'); ?></label>
+				<div class="input-group">
+					<input
+						name="passwd"
```
Comment (brianteeman) [inline administrator/modules/mod_login/tmpl/default.php]: please order name, type

File: administrator/modules/mod_login/tmpl/default.php [hunk 3]
```diff
@@ -42,21 +44,23 @@ class="form-control input-full"
...
+						id="mod-login-password"
+						type="password"
+						class="form-control input-full"
+						required="required"
+	                    tabindex="2"
```
Comment (brianteeman) [inline administrator/modules/mod_login/tmpl/default.php:55]: tab s on this and the next line

File: administrator/language/en-GB/en-GB.mod_login.ini (hunk 1 above)
Comment (dgrammatiko) [reply administrator/language/en-GB/en-GB.mod_login.ini]: This is just for easy testing, but should be removed now

File: administrator/modules/mod_login/tmpl/default.php [hunk 4]
```diff
@@ -42,21 +38,10 @@ class="form-control input-full"
...
-					class="form-control input-full"
-					required="required"
-                    tabindex="2"
-                >
-				<span class="input-group-addon">
-					<span class="fa fa-eye" aria-hidden="true"></span>
-					<span class="sr-only"><?php echo Text::_('JSHOW'); ?></span>
-				</span>
-			</div>
+			<joomla-field-password reveal="true" min-length="2" text-show="<?php echo Text::_('JSHOW'); ?>" text-hide="<?php echo Text::_('JSHOW'); ?>">
```
Comment (Quy) [inline administrator/modules/mod_login/tmpl/default.php]: For `text-hide` attribute, should it be `JFIELD_PASSWORD_INDICATE_COMPLETE`?

Comment (dgrammatiko) [reply administrator/modules/mod_login/tmpl/default.php]: JHIDE, my bad

File: media/system/webcomponents/js/joomla-field-password.js [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+					document.formvalidator.setHandler(this.getAttribute('validation-handler'), this.handler.bind(this));
+				}
+			}
+
+			if (this.reveal === 'true') {
+				const parent = document.createElement('span');
+				const firstSpan = document.createElement('span');
+				const secondSpan = document.createElement('span');
+
+				parent.classList.add('input-group-addon');
```
Comment (C-Lodder) [inline media/system/webcomponents/js/joomla-field-password.js]: This needs changing to `parent.classList.add('input-group-append');` and then needs another child element inside with the class `input-group-text`

```
<span class="input-group-append">
  <span class="input-group-text">
    <span class="fa fa-eye" aria-hidden="true"></span>
  </span>
</span>
```

Comment (dgrammatiko) [reply media/system/webcomponents/js/joomla-field-password.js]: No, that’s bootstrap specific. Custom elements are framework agnostic!

Comment (C-Lodder) [reply media/system/webcomponents/js/joomla-field-password.js]: In which case let's add the markup to the field rather than adding it via JS?

Comment (dgrammatiko) [reply media/system/webcomponents/js/joomla-field-password.js]: We should make the least possible html markup, so even dummies can play well with the elements. As I wrote in the description this needs still some work:
- remove bs dependency
- remove font awesome dependency
- finilise the custom validation handler method (this starts here but will impact all input elements)

All and all this is still WIP, feel free to pr anything you feel should be done

Comment (C-Lodder) [reply media/system/webcomponents/js/joomla-field-password.js]: Here are the SVG's for the eyes:

https://github.com/encharm/Font-Awesome-SVG-PNG/blob/master/black/svg/eye-slash.svg
https://github.com/encharm/Font-Awesome-SVG-PNG/blob/master/black/svg/eye.svg

Comment (dgrammatiko) [reply media/system/webcomponents/js/joomla-field-password.js]: I setup a codepen so we can work the html/css part easier: https://codepen.io/dgt41/pen/mpwjZN

**Comment by dgrammatiko** [comment]
The Todo list is still relevant:

- Still Bootstrap specific
- Still Font awesome specific
- Needs A11Y audit
- Needs final approval form the design team

**Comment by dgrammatiko** [comment]
BTW the show/hide button is rendered properly now:
screen shot 2018-03-23 at 19 35 18
screen shot 2018-03-23 at 19 38 10

**Comment by brianteeman** [comment]
@dgrammatiko can you update this to resolve the conflicts so that this can be tested please

**Comment by dgrammatiko** [comment]
sure

PS  there are still a few todo that needs to be done at some point (preferably before merging) 
https://github.com/joomla/joomla-cms/pull/19231#issuecomment-375764358

**Comment by brianteeman** [comment]
Tested on the new user page in the admin

1. This page has two password fields - which produces two copies of the css to be inserted. Surely thats not correct
chrome_2018-07-22_20-16-32

2. There appears to be a missing close > which is stopping it working
chrome_2018-07-22_20-17-59

**Comment by dgrammatiko** [comment]
Ok I've messed up the sync, will fix it

**Comment by brianteeman** [comment]
As a temp measure until you resolve the outstanding issues and this can be accepted I did a quick pr to fix the button styling on the current code #21462

**Pull request closed without merging**